console/v1/types_console_sample.go

```diff
@@ -0,0 +1,183 @@
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ConsoleSample is an extension to customizing OpenShift web console by adding samples.
+//
+// Compatibility level 2: Stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=2
+type ConsoleSample struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata"`
+
+	// +required
+	Spec ConsoleSampleSpec `json:"spec"`
+}
+
+// ConsoleSampleSpec is the desired sample for the web console.
+// Samples will appear with their title, descriptions and a badge in a samples catalog.
+type ConsoleSampleSpec struct {
+	// title is the display name of the sample.
+	//
+	// It is required and must be no more than 50 characters in length.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=50
+	Title string `json:"title"`
+
+	// abstract is a short teaser the sample.
+	//
+	// It is required and must be no more than 100 characters in length.
+	//
+	// Currently, the abstract is shown on the sample card tile below the title and provider
+	// and is limited to three lines of content.
+	// +required
+	// +kubebuilder:validation:MaxLength=100
+	Abstract string `json:"abstract"`
+
+	// description of the sample.
+	//
+	// It is required and can have a maximum length of **4096** characters.
+	//
+	// It is a README.md-like content for additional information, links, pre-conditions, and other instructions.
+	// It will be rendered as Markdown so that it can contain line breaks, links, and other simple formatting.
+	// +required
+	// +kubebuilder:validation:MaxLength=4096
+	Description string `json:"description"`
+
+	// icon is an optional base64 encoded image and shown beside the sample title.
+	//
+	// The format must follow the [data: URL format].
+	//
+	//   data:[<mediatype>][;base64],<base64 encoded image>
+	//
+	// For example:
+	//
+	//   data:image;base64,             plus the base64 encoded image.
+	//
+	// Vector images can also be used. SVG icons must start with:
+	//
+	//   data:image/svg+xml;base64,     plus the base64 encoded SVG image.
+	//
+	// All sample catalog icons will be shown on a white background (also when the dark theme is used).
+	// The web console ensures that different aspect radios work correctly.
+	// Currently, the surface of the icon is at most 40x100px.
+	//
+	// [data: URL format]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs
+	// +optional
+	// +kubebuilder:validation:Pattern=`^data:`
```

Comment:
The pattern should always include `base64,` right? Can we update the regex?
I had a play with `^data:([a-z/\.+0-9]*;(([-a-zA-Z0-9=])*;)?)?base64,` and think that could work reviewing the content in https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs

Reply:
`^data:([a-z/\.+0-9]*;(([-a-zA-Z0-9=])*;)?)?base64,` would work for the testcases I would test.

```
data:image/svg;base64,asd
data:image/svg+xml;base64,asd
data:image/png;base64,asd
data:text/plain;charset=US-ASCII;base64,asd
```

I will test it on a cluster.